# Training the model

In [6]:
import datareader as dr
import word2vec as wv
import time
import util

In [3]:
ts = time.time()
file_path = "./data/pt96.txt"
my_data = dr.DataReader(file_path)
my_data.get_data()
word2index = my_data.word2index
index2word = my_data.index2word
config = wv.Config()
my_model = wv.SkipGramModel(config)
embeddings = wv.run_training(my_model, my_data)
te = time.time()

Initialized

&&&&&&&&& For TensorBoard visualization type &&&&&&&&&&&

tensorboard  --logdir=./graphs/25-03-2017_12-25-21


&&&&&&&&& And for the 3d embedding visualization type &&

tensorboard  --logdir=./processed

Nearest to três: hífen, visconde, Kunisada, vazaram, Brasil3, largos, defensivas, código,
Nearest to maior: vencida, Konder, Marvel, Asuma, UK, Avatar, magnatas, abril,
Nearest to a: furacão, mortos, Carrier, congelada, tu, construir, ceia, conhecerem,
Nearest to dia: união, natural, negavam, complexo, lona, 1777, pote, Messias,
Nearest to é: açougue, Faraó, estado, Shukaku, Jia, Tour, Shizune, perdiam,
Nearest to de: Nightmare, coescrita, reabilitar, heterodoxas, utilizarem, confiscados, manganês, maneirista,
Nearest to ano: Trácia, divertida, 2099, derrubar, poderei, Anatomy, Calcedônia, interferiu,
Nearest to durante: navios, Yes, Stirling, Hard, primeiro, extremas, Elfos, picada,
Nearest to à: pisar, Speedvision, pecuária, Santi, ovinos, Landing, Apiaceae, golfinho,
Ne

In [ ]:
#! tensorboard ..

In [55]:
#! tensorboard  --logdir=./processed

Starting TensorBoard b'41' on port 6006
(You can navigate to http://127.0.1.1:6006)
^CTraceback (most recent call last):
  File "/usr/local/bin/tensorboard", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/tensorboard/tensorboard.py", line 151, in main
    tb_server.serve_forever()
  File "/usr/lib/python3.5/socketserver.py", line 232, in serve_forever
    ready = selector.select(poll_interval)
  File "/usr/lib/python3.5/selectors.py", line 376, in select
    fd_event_list = self._poll.poll(timeout)
KeyboardInterrupt



In [9]:
print("Duration of the training = {}".format(te-ts))

Duration of the training = 612.1953608989716


# Evaluating the model

In [52]:
eval_path = "./evaluation/questions-words-ptbr.txt"
score, report = util.score(index2word, word2index, embeddings, eval_path,verbose=False,raw=True)



: capital-common-countries



: capital-world



: currency



: city-in-state



: family



: gram1-adjective-to-adverb



: gram2-opposite



: gram3-comparative



: gram4-superlative



: gram5-present-participle



: gram6-nationality-adjective



: gram7-past-tense



: gram8-plural



: gram9-plural-verbs




In [53]:
print("Overall score ={}".format(score))
print("Detailed score:")
for result in report:
    print(result)

Overall score =0.007391763463569166
Detailed score:
capital-common-countries: (0/306)
capital-world: (0/1155)
currency: (0/106)
city-in-state: (0/1169)
family: (40/342)
gram1-adjective-to-adverb: (0/552)
gram2-opposite: (0/182)
gram3-comparative: (7/30)
gram4-superlative: (3/20)
gram5-present-participle: (2/702)
gram6-nationality-adjective: (0/412)
gram7-past-tense: (0/1056)
gram8-plural: (0/992)
gram9-plural-verbs: (4/552)


# Comparing with gensim

gensim.models.Word2Vec(sentences=None, size=100, alpha=0.025, window=5, min_count=5, max_vocab_size=None, sample=0.001, seed=1, workers=3, min_alpha=0.0001, sg=0, hs=0, negative=5, cbow_mean=1, hashfxn=<built-in function hash>, iter=5, null_word=0, trim_rule=None, sorted_vocab=1, batch_words=10000)

In [14]:
import gensim, logging
import os
import inspect
logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)


class MySentences(object):
    def __init__(self, dirname):
        self.dirname = dirname
 
    def __iter__(self):
        for fname in os.listdir(self.dirname):
            for line in open(os.path.join(self.dirname, fname)):
                yield line.split()

file_path = "./data/gensim"
sentences = MySentences(file_path)
model = gensim.models.Word2Vec(sentences)

2017-03-25 12:38:03,713 : INFO : collecting all words and their counts
2017-03-25 12:38:03,714 : INFO : PROGRESS: at sentence #0, processed 0 words, keeping 0 word types
2017-03-25 12:38:03,745 : INFO : PROGRESS: at sentence #10000, processed 57554 words, keeping 12910 word types
2017-03-25 12:38:03,778 : INFO : PROGRESS: at sentence #20000, processed 136777 words, keeping 25063 word types
2017-03-25 12:38:03,815 : INFO : PROGRESS: at sentence #30000, processed 228679 words, keeping 37221 word types
2017-03-25 12:38:03,849 : INFO : PROGRESS: at sentence #40000, processed 310436 words, keeping 45452 word types
2017-03-25 12:38:03,878 : INFO : PROGRESS: at sentence #50000, processed 371745 words, keeping 49804 word types
2017-03-25 12:38:03,903 : INFO : PROGRESS: at sentence #60000, processed 430630 words, keeping 53594 word types
2017-03-25 12:38:03,932 : INFO : PROGRESS: at sentence #70000, processed 490872 words, keeping 57220 word types
2017-03-25 12:38:03,953 : INFO : PROGRESS: at s

In [15]:
model.accuracy(eval_path)

2017-03-25 12:39:42,123 : INFO : precomputing L2-norms of word weight vectors
2017-03-25 12:39:42,552 : INFO : capital-common-countries: 8.9% (8/90)
2017-03-25 12:39:43,247 : INFO : capital-world: 2.3% (4/173)
2017-03-25 12:39:43,477 : INFO : currency: 0.0% (0/54)
2017-03-25 12:39:44,294 : INFO : city-in-state: 0.5% (1/208)
2017-03-25 12:39:45,458 : INFO : family: 43.1% (132/306)
2017-03-25 12:39:46,922 : INFO : gram1-adjective-to-adverb: 1.3% (5/380)
2017-03-25 12:39:47,297 : INFO : gram2-opposite: 3.3% (3/90)
2017-03-25 12:39:47,431 : INFO : gram3-comparative: 40.0% (12/30)
2017-03-25 12:39:47,477 : INFO : gram4-superlative: 66.7% (4/6)
2017-03-25 12:39:49,240 : INFO : gram5-present-participle: 13.2% (61/462)
2017-03-25 12:39:52,175 : INFO : gram6-nationality-adjective: 5.0% (37/739)
2017-03-25 12:39:54,342 : INFO : gram7-past-tense: 24.5% (124/506)
2017-03-25 12:39:56,190 : INFO : gram8-plural: 6.6% (25/380)
2017-03-25 12:39:58,109 : INFO : gram9-plural-verbs: 9.5% (29/306)
2017-03-

[{'correct': [('PEQUIM', 'CHINA', 'MOSCOU', 'RÚSSIA'),
   ('BERLIM', 'ALEMANHA', 'LONDRES', 'INGLATERRA'),
   ('BERLIM', 'ALEMANHA', 'MOSCOU', 'RÚSSIA'),
   ('LONDRES', 'INGLATERRA', 'PARIS', 'FRANÇA'),
   ('MOSCOU', 'RÚSSIA', 'ROMA', 'ITÁLIA'),
   ('PARIS', 'FRANÇA', 'ATENAS', 'GRÉCIA'),
   ('PARIS', 'FRANÇA', 'MOSCOU', 'RÚSSIA'),
   ('ESTOCOLMO', 'SUÉCIA', 'PARIS', 'FRANÇA')],
  'incorrect': [('ATENAS', 'GRÉCIA', 'PEQUIM', 'CHINA'),
   ('ATENAS', 'GRÉCIA', 'BERLIM', 'ALEMANHA'),
   ('ATENAS', 'GRÉCIA', 'LONDRES', 'INGLATERRA'),
   ('ATENAS', 'GRÉCIA', 'MADRID', 'ESPANHA'),
   ('ATENAS', 'GRÉCIA', 'MOSCOU', 'RÚSSIA'),
   ('ATENAS', 'GRÉCIA', 'PARIS', 'FRANÇA'),
   ('ATENAS', 'GRÉCIA', 'ROMA', 'ITÁLIA'),
   ('ATENAS', 'GRÉCIA', 'ESTOCOLMO', 'SUÉCIA'),
   ('ATENAS', 'GRÉCIA', 'TÓQUIO', 'JAPÃO'),
   ('PEQUIM', 'CHINA', 'BERLIM', 'ALEMANHA'),
   ('PEQUIM', 'CHINA', 'LONDRES', 'INGLATERRA'),
   ('PEQUIM', 'CHINA', 'MADRID', 'ESPANHA'),
   ('PEQUIM', 'CHINA', 'PARIS', 'FRANÇA'),
   ('PEQUIM

### Gensim
capital-common-countries: 8.9% (8/90)

capital-world: 2.3% (4/173)

currency: 0.0% (0/54)

city-in-state: 0.5% (1/208)

family: 43.1% (132/306)

gram1-adjective-to-adverb: 1.3% (5/380)

gram2-opposite: 3.3% (3/90)

gram3-comparative: 40.0% (12/30)

gram4-superlative: 66.7% (4/6)

gram5-present-participle: 13.2% (61/462)

gram6-nationality-adjective: 5.0% (37/739)

gram7-past-tense: 24.5% (124/506)

gram8-plural: 6.6% (25/380)

gram9-plural-verbs: 9.5% (29/306)

total: 11.9% (445/3730)


### Me

capital-common-countries: (0/306)

capital-world: (0/1155)

currency: (0/106)

city-in-state: (0/1169)

family: (40/342)

gram1-adjective-to-adverb: (0/552)

gram2-opposite: (0/182)

gram3-comparative: (7/30)

gram4-superlative: (3/20)

gram5-present-participle: (2/702)

gram6-nationality-adjective: (0/412)

gram7-past-tense: (0/1056)

gram8-plural: (0/992)

gram9-plural-verbs: (4/552)
